sno/dataset2.py

```diff
@@ -122,6 +125,141 @@ def hexhash(self):
         return _hexhash(self.dumps())
 
 
+def _pk_index_ordering(column):
+    """Returns primary key columns first, in pk_index order, then other columns."""
+    if column.pk_index is not None:
+        return column.pk_index
+    else:
+        # Non primary-key columns after primary key columns.
+        return float('inf')
+
+
+class ColumnSchema:
+    """
+    The schema for a single column. A column has a unique ID that is constant for the columns lifetime
+    - even if the column is moved or renamed - and it has a name, a datatype, info about if this
+    column is one of the primary keys, and potentially extra info about the specific datatype.
+    """
+
+    def __init__(self, id, name, data_type, pk_index=None, extra_type_info=None):
```

Comment:
```suggestion
    def __init__(self, id, name, data_type, pk_index=None, **extra_type_info):
```
might be simpler? Simplifies `from_json_dict()` too.

Reply:
Done